pkg/controller/cloneset/sync/cloneset_scale.go

```diff
@@ -403,3 +420,101 @@ func (r *realControl) choosePodsToDelete(cs *appsv1alpha1.CloneSet, totalDiff in
 
 	return podsToDelete
 }
+
+func getInstanceIDsFromPods(pods []*v1.Pod) sets.String {
+	ins := sets.NewString()
+	for _, pod := range pods {
+		if id := clonesetutils.GetInstanceID(pod); id != "" {
+			ins.Insert(id)
+		}
+	}
+	return ins
+}
+
+func getUselessPVCs(pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim) map[string]*v1.PersistentVolumeClaim {
+	activeIDs := getInstanceIDsFromPods(pods)
+
+	useless := map[string]*v1.PersistentVolumeClaim{}
+	for _, pvc := range pvcs {
+		id := clonesetutils.GetInstanceID(pvc)
+		if id != "" && !activeIDs.Has(id) {
+			useless[id] = pvc
+		}
+	}
+	return useless
+}
+
+func (r *realControl) cleanupPVCs(cs *appsv1alpha1.CloneSet, uselessPVCs map[string]*v1.PersistentVolumeClaim) (bool, error) {
+	var modified bool
+
+	pods, err := getAllPods(r.Client, cs)
+	if err != nil {
+		klog.Errorf("Could not get cloneset %s owned pods", clonesetutils.GetControllerKey(cs))
+		return modified, err
+	}
+
+	// If useless pvc owner pod does not exist, the pvc can be deleted directly,
+	// else update pvc's ownerreference to pod.
+	for _, pod := range pods {
```

Comment:
can we really find a match pod for "useless" pvc? if we can is the pvc really useless in the begining

Reply:
"useless" here means that the pvc can be deleted finally, e.g. it's owner pod is stuck in terminating state, the pvc should be deleted after the pod is cleaned.